File: client/src/lib/sanitization.ts
export interface SanitizationOptions {
  maxLength?: number
  allowHtml?: boolean
  allowedTags?: string[]
  stripNewlines?: boolean
  trimWhitespace?: boolean
}

export class InputSanitizer {
  // Default maximum lengths for different field types
  static readonly MAX_LENGTHS = {
    title: 200,
    message: 2000,
    description: 5000,
    pros: 1000,
    cons: 1000,
    stepsToReproduce: 2000,
    expectedResult: 1000,
    actualResult: 1000,
    suggestedSolution: 2000,
    benefits: 1000,
    useCase: 1000,
    comment: 1000,
    submitterName: 100,
    submitterEmail: 254,
    category: 50,
    priority: 20,
    severity: 20,
    general: 1000,
  } as const

  // HTML tags that are considered safe
  static readonly SAFE_HTML_TAGS = [
    'b',
    'strong',
    'i',
    'em',
    'u',
    'br',
    'p',
    'ul',
    'ol',
    'li',
    'code',
    'pre',
  ]

  /**
   * Sanitize text input by removing dangerous content and applying formatting rules
   */
  static sanitizeText(input: string | null | undefined, options: SanitizationOptions = {}): string {
    if (!input) return ''

    const {
      maxLength = this.MAX_LENGTHS.general,
      allowHtml = false,
      allowedTags = this.SAFE_HTML_TAGS,
      stripNewlines = false,
      trimWhitespace = true,
    } = options

    let sanitized = String(input)

    // Trim whitespace
    if (trimWhitespace) {
      sanitized = sanitized.trim()
    }

    // Remove null bytes and control characters (except newlines and tabs)
    sanitized = sanitized
      .split('')
      .filter((char) => {
        const code = char.charCodeAt(0)
        return code !== 0 && (code < 9 || code > 13) && code !== 127
      })
      .join('')

    // Handle newlines
    if (stripNewlines) {
      sanitized = sanitized.replace(/[\r\n]/g, ' ')
    } else {
      // Normalize newlines
      sanitized = sanitized.replace(/\r\n/g, '\n').replace(/\r/g, '\n')
    }

    // HTML sanitization
    if (allowHtml) {
      sanitized = this._sanitizeHtml(sanitized, allowedTags)
    } else {
      // Escape HTML to prevent XSS
      sanitized = this._escapeHtml(sanitized)
    }

    // Limit length
    if (maxLength && sanitized.length > maxLength) {
      sanitized = sanitized.substring(0, maxLength)
    }

    return sanitized
  }

  /**
   * Sanitize email address
   */
  static sanitizeEmail(email: string | null | undefined): string | null {
    if (!email) return null

    const sanitized = email.trim().toLowerCase()

    // Basic email validation pattern
    const emailPattern = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/

    if (!emailPattern.test(sanitized)) {
      return null
    }

    // Limit length
    return sanitized.length > 254 ? null : sanitized
  }

  /**
   * Sanitize URL input
   */
  static sanitizeUrl(url: string | null | undefined): string | null {
    if (!url) return null

    let sanitized = url.trim()

    // Add protocol if missing
    if (!sanitized.startsWith('http://') && !sanitized.startsWith('https://')) {
      sanitized = 'https://' + sanitized
    }

    try {
      // Basic URL validation
      new URL(sanitized)

      // Limit length
      return sanitized.length > 2048 ? null : sanitized
    } catch {
      return null
    }
  }

  /**
   * Sanitize rating values
   */
  static sanitizeRating(rating: number | string | null | undefined): number | null {
    if (rating === null || rating === undefined) return null

    try {
      const num = Number(rating)
      if (isNaN(num)) return null

      const intRating = Math.round(num)

      // Ensure rating is within valid range (0-10)
      if (intRating >= 0 && intRating <= 10) {
        return intRating
      }
    } catch {
      // Ignore errors
    }

    return null
  }

  /**
   * Sanitize severity values
   */
  static sanitizeSeverity(severity: string | null | undefined): string | null {
    if (!severity) return null

    const sanitized = severity.trim().toLowerCase()
    const allowedValues = ['low', 'medium', 'high', 'critical']

    return allowedValues.includes(sanitized) ? sanitized : 'medium'
  }

  /**
   * Sanitize priority values
   */
  static sanitizePriority(priority: string | null | undefined): string | null {
    if (!priority) return null

    const sanitized = priority.trim().toLowerCase()
    const allowedValues = ['low', 'medium', 'high', 'urgent']

    return allowedValues.includes(sanitized) ? sanitized : 'medium'
  }

  /**
   * Sanitize category values
   */
  static sanitizeCategory(category: string | null | undefined): string | null {
    if (!category) return null

    const sanitized = category.trim()

    // Remove special characters and limit length
    const cleaned = sanitized.replace(/[^\w\s-]/g, '')
    const limited = cleaned.substring(0, 50)

    return limited || null
  }

  /**
   * Sanitize widget key
   */
  static sanitizeWidgetKey(key: string | null | undefined): string | null {
    if (!key) return null

    const sanitized = key.trim()

    // Widget keys should only contain alphanumeric characters and underscores
    if (/^[a-zA-Z0-9_]+$/.test(sanitized)) {
      return sanitized
    }

    return null
  }

  /**
   * Get character count for input validation
   */
  static getCharacterCount(input: string | null | undefined): number {
    return input ? input.length : 0
  }

  /**
   * Check if input exceeds maximum length
   */
  static isExceedingMaxLength(input: string | null | undefined, maxLength: number): boolean {
    return this.getCharacterCount(input) > maxLength
  }

  /**
   * Get remaining characters
   */
  static getRemainingCharacters(input: string | null | undefined, maxLength: number): number {
    return Math.max(0, maxLength - this.getCharacterCount(input))
  }

  /**
   * Private method to escape HTML
   */
  private static _escapeHtml(text: string): string {
    const div = document.createElement('div')
    div.textContent = text
    return div.innerHTML
  }

  /**
   * Private method to sanitize HTML content
   */
  private static _sanitizeHtml(html: string, allowedTags: string[]): string {
    // This is a simplified approach - in production, consider using DOMPurify
    const div = document.createElement('div')
    div.innerHTML = html

    // Remove all script tags and event handlers
    const scripts = div.querySelectorAll('script')
    scripts.forEach((script) => script.remove())

    // Remove all elements except allowed tags
    const allElements = div.querySelectorAll('*')
    allElements.forEach((element) => {
      if (!allowedTags.includes(element.tagName.toLowerCase())) {
        element.replaceWith(element.textContent || '')
      }

      // Remove all attributes
      Array.from(element.attributes).forEach((attr) => {
        element.removeAttribute(attr.name)
      })
    })

    return div.innerHTML
  }
}

/**
 * Convenience functions for common sanitization tasks
 */
export const sanitizeInput = {
  title: (input: string) =>
    InputSanitizer.sanitizeText(input, { maxLength: InputSanitizer.MAX_LENGTHS.title }),
  message: (input: string) =>
    InputSanitizer.sanitizeText(input, { maxLength: InputSanitizer.MAX_LENGTHS.message }),
  description: (input: string) =>
    InputSanitizer.sanitizeText(input, { maxLength: InputSanitizer.MAX_LENGTHS.description }),
  comment: (input: string) =>
    InputSanitizer.sanitizeText(input, { maxLength: InputSanitizer.MAX_LENGTHS.comment }),
  email: (input: string) => InputSanitizer.sanitizeEmail(input),
  url: (input: string) => InputSanitizer.sanitizeUrl(input),
  rating: (input: number | string) => InputSanitizer.sanitizeRating(input),
  severity: (input: string) => InputSanitizer.sanitizeSeverity(input),
  priority: (input: string) => InputSanitizer.sanitizePriority(input),
  category: (input: string) => InputSanitizer.sanitizeCategory(input),
  widgetKey: (input: string) => InputSanitizer.sanitizeWidgetKey(input),
  general: (input: string, maxLength?: number) =>
    InputSanitizer.sanitizeText(input, {
      maxLength: maxLength || InputSanitizer.MAX_LENGTHS.general,
    }),
}

/**
 * React hook for input sanitization (if you want to use it as a hook)
 */
export const useInputSanitization = (maxLength?: number) => {
  const sanitize = (input: string) => InputSanitizer.sanitizeText(input, { maxLength })
  const getCharCount = (input: string) => InputSanitizer.getCharacterCount(input)
  const isExceeding = (input: string) =>
    InputSanitizer.isExceedingMaxLength(input, maxLength || InputSanitizer.MAX_LENGTHS.general)
  const getRemaining = (input: string) =>
    InputSanitizer.getRemainingCharacters(input, maxLength || InputSanitizer.MAX_LENGTHS.general)

  return {
    sanitize,
    getCharCount,
    isExceeding,
    getRemaining,
    maxLength: maxLength || InputSanitizer.MAX_LENGTHS.general,
  }
}

File: api/server/app/core/sanitization.py
import re
import html
from typing import Any, Dict, Optional, Union


class InputSanitizer:
    ALLOWED_HTML_TAGS = {
        'b',
        'strong',
        'i',
        'em',
        'u',
        'br',
        'p',
        'ul',
        'ol',
        'li',
        'code',
        'pre',
    }

    MAX_LENGTHS = {
        'title': 200,
        'message': 2000,
        'description': 5000,
        'pros': 1000,
        'cons': 1000,
        'steps_to_reproduce': 2000,
        'expected_result': 1000,
        'actual_result': 1000,
        'suggested_solution': 2000,
        'benefits': 1000,
        'use_case': 1000,
        'comment': 1000,
        'submitter_name': 100,
        'submitter_email': 254,
        'category': 50,
        'priority': 20,
        'severity': 20,
    }

    @classmethod
    def sanitize_text(
        cls, text: Optional[str], max_length: Optional[int] = None
    ) -> Optional[str]:
        if not text:
            return text
        text = str(text).strip()
        text = re.sub(r'[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]', '', text)
        text = html.escape(text)
        text = cls._restore_safe_html(text)

        if max_length:
            text = text[:max_length]

        return text

    @classmethod
    def _restore_safe_html(cls, text: str) -> str:
        safe_patterns = [
            (r'&lt;b&gt;', '<b>'),
            (r'&lt;/b&gt;', '</b>'),
            (r'&lt;strong&gt;', '<strong>'),
            (r'&lt;/strong&gt;', '</strong>'),
            (r'&lt;i&gt;', '<i>'),
            (r'&lt;/i&gt;', '</i>'),
            (r'&lt;em&gt;', '<em>'),
            (r'&lt;/em&gt;', '</em>'),
            (r'&lt;u&gt;', '<u>'),
            (r'&lt;/u&gt;', '</u>'),
            (r'&lt;br&gt;', '<br>'),
            (r'&lt;p&gt;', '<p>'),
            (r'&lt;/p&gt;', '</p>'),
            (r'&lt;ul&gt;', '<ul>'),
            (r'&lt;/ul&gt;', '</ul>'),
            (r'&lt;ol&gt;', '<ol>'),
            (r'&lt;/ol&gt;', '</ol>'),
            (r'&lt;li&gt;', '<li>'),
            (r'&lt;/li&gt;', '</li>'),
            (r'&lt;code&gt;', '<code>'),
            (r'&lt;/code&gt;', '</code>'),
            (r'&lt;pre&gt;', '<pre>'),
            (r'&lt;/pre&gt;', '</pre>'),
        ]

        for pattern, replacement in safe_patterns:
            text = re.sub(pattern, replacement, text, flags=re.IGNORECASE)

        return text

    @classmethod
    def sanitize_email(cls, email: Optional[str]) -> Optional[str]:
        if not email:
            return email

        email = str(email).strip().lower()

        email_pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'

        if not re.match(email_pattern, email):
            return None

        return email[:254]

    @classmethod
    def sanitize_url(cls, url: Optional[str]) -> Optional[str]:
        if not url:
            return url

        url = str(url).strip()

        if not url.startswith(('http://', 'https://')):
            url = 'https://' + url

        try:
            if len(url) > 2048:
                return None
            return url
        except Exception:
            return None

    @classmethod
    def sanitize_rating(cls, rating: Optional[Union[int, float]]) -> Optional[int]:
        if rating is None:
            return None

        try:
            rating = int(float(rating))

            if 1 <= rating <= 10:
                return rating
        except (ValueError, TypeError):
            pass

        return None

    @classmethod
    def sanitize_severity(cls, severity: Optional[str]) -> Optional[str]:
        if not severity:
            return None

        severity = str(severity).strip().lower()
        allowed_values = {'low', 'medium', 'high', 'critical'}

        return severity if severity in allowed_values else 'medium'

    @classmethod
    def sanitize_priority(cls, priority: Optional[str]) -> Optional[str]:
        if not priority:
            return None

        priority = str(priority).strip().lower()
        allowed_values = {'low', 'medium', 'high', 'urgent'}

        return priority if priority in allowed_values else 'medium'

    @classmethod
    def sanitize_category(cls, category: Optional[str]) -> Optional[str]:
        if not category:
            return None

        category = str(category).strip()

        category = re.sub(r'[^\w\s-]', '', category)
        category = category[:50]

        return category if category else None

    @classmethod
    def sanitize_feedback_data(cls, data: Dict[str, Any]) -> Dict[str, Any]:
        sanitized = {}

        for field, value in data.items():
            if value is None:
                continue

            if field in cls.MAX_LENGTHS:
                max_length = cls.MAX_LENGTHS[field]

                if field == 'submitter_email':
                    sanitized[field] = cls.sanitize_email(value)
                elif field in [
                    'score',
                    'overall_rating',
                    'nps_score',
                    'csat_score',
                    'ces_score',
                ]:
                    sanitized[field] = cls.sanitize_rating(value)
                elif field == 'severity':
                    sanitized[field] = cls.sanitize_severity(value)
                elif field == 'priority':
                    sanitized[field] = cls.sanitize_priority(value)
                elif field == 'category':
                    sanitized[field] = cls.sanitize_category(value)
                elif field in ['visual_proof', 'context'] and isinstance(value, dict):
                    sanitized[field] = cls.sanitize_feedback_data(value)
                else:
                    sanitized[field] = cls.sanitize_text(value, max_length)
            else:
                if isinstance(value, str):
                    sanitized[field] = cls.sanitize_text(value, 1000)
                elif isinstance(value, dict):
                    sanitized[field] = cls.sanitize_feedback_data(value)
                else:
                    sanitized[field] = value

        return sanitized

    @classmethod
    def sanitize_widget_key(cls, widget_key: Optional[str]) -> Optional[str]:
        if not widget_key:
            return None

        widget_key = str(widget_key).strip()

        if re.match(r'^[a-zA-Z0-9_]+$', widget_key):
            return widget_key

        return None
